federation/pkg/federation-controller/job/jobcontroller.go

```diff
@@ -0,0 +1,538 @@
+			fjc.clusterDeliverer.DeliverAfter(allClustersKey, nil, clusterUnavailableDelay)
+		},
+	}
+	fjc.fedJobInformer = fedutil.NewFederatedInformer(fedClient, jobFedInformerFactory, &clusterLifecycle)
+
+	fjc.jobStore, fjc.jobController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				versionedOptions := fedutil.VersionizeV1ListOptions(options)
+				return fjc.fedClient.Batch().Jobs(apiv1.NamespaceAll).List(versionedOptions)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				versionedOptions := fedutil.VersionizeV1ListOptions(options)
+				return fjc.fedClient.Batch().Jobs(apiv1.NamespaceAll).Watch(versionedOptions)
+			},
+		},
+		&batchv1.Job{},
+		controller.NoResyncPeriodFunc(),
+		fedutil.NewTriggerOnMetaAndSpecChanges(
+			func(obj runtime.Object) { fjc.deliverFedJobObj(obj, 0) },
+		),
+	)
+
+	fjc.fedUpdater = fedutil.NewFederatedUpdater(fjc.fedJobInformer,
+		func(client kubeclientset.Interface, obj runtime.Object) error {
+			rs := obj.(*batchv1.Job)
+			_, err := client.Batch().Jobs(rs.Namespace).Create(rs)
+			return err
+		},
+		func(client kubeclientset.Interface, obj runtime.Object) error {
+			rs := obj.(*batchv1.Job)
+			_, err := client.Batch().Jobs(rs.Namespace).Update(rs)
+			return err
+		},
+		func(client kubeclientset.Interface, obj runtime.Object) error {
+			rs := obj.(*batchv1.Job)
+			err := client.Batch().Jobs(rs.Namespace).Delete(rs.Name, &apiv1.DeleteOptions{})
+			return err
+		})
+
+	return fjc
+}
+
+func (fjc *JobController) Run(workers int, stopCh <-chan struct{}) {
+	go fjc.jobController.Run(stopCh)
+	fjc.fedJobInformer.Start()
+
+	fjc.jobDeliverer.StartWithHandler(func(item *fedutil.DelayingDelivererItem) {
+		fjc.jobWorkQueue.Add(item.Key)
+	})
+	fjc.clusterDeliverer.StartWithHandler(func(_ *fedutil.DelayingDelivererItem) {
+		fjc.reconcileJobsOnClusterChange()
+	})
+
+	for !fjc.isSynced() {
+		time.Sleep(5 * time.Millisecond)
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(fjc.worker, time.Second, stopCh)
+	}
+
+	fedutil.StartBackoffGC(fjc.jobBackoff, stopCh)
+
+	<-stopCh
+	glog.Infof("Shutting down JobController")
+	fjc.jobDeliverer.Stop()
+	fjc.clusterDeliverer.Stop()
+	fjc.jobWorkQueue.ShutDown()
+	fjc.fedJobInformer.Stop()
+}
+
+func (fjc *JobController) isSynced() bool {
+	if !fjc.fedJobInformer.ClustersSynced() {
+		glog.Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := fjc.fedJobInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !fjc.fedJobInformer.GetTargetStore().ClustersSynced(clusters) {
+		glog.V(2).Infof("cluster job list not synced")
+		return false
+	}
+
+	if !fjc.jobController.HasSynced() {
+		glog.V(2).Infof("federation job list not synced")
+		return false
+	}
+	return true
+}
+
+func (fjc *JobController) deliverLocalJob(obj interface{}, duration time.Duration) {
```

Comment:
the behavior is different from deliverFedJob in order to filter out those created directly upon local (underlying) clusters which shouldn't trigger an reconciliation.